# ESCUELA POLITECNICA NACIONAL
# FACULTAD DE INGENIERIA EN SISTEMAS
# METODOS NUMERICOS

Nombre: Peñafiel Jácome Melanie Dayana


Fecha: 13 / 05 / 2026

--------

# Tarea 03 - Algoritmos y Complejidad

------

# Ejercicio 1

Utilice aritmética de corte de tres dígitos para calcular las siguientes sumas. Para cada parte, ¿qué método es 
más preciso y por qué? 

a. ∑ (1 𝑖2⁄ ) 10
𝑖=1 primero por 1
1
+1
4
+⋯+ 1
100
 y luego por 1
100
+1
81
+⋯+1
1
 
 b. ∑ (1 𝑖3⁄ ) 10
𝑖=1 primero por 1
1
+1
8
+1
27
+⋯+ 1
1000
y luego por 1
1000
+1/ 729 +⋯+ 1/1


In [1]:
import math

def cortardigitos(x):                     
    if x == 0: 
        return 0.0
    exp = math.floor(math.log10(abs(x)))        
    factor = 10 ** (2 - exp)                    
    return math.trunc(x * factor) / factor      

def sumacorte(serie, orden='asc'):                                
    total = 0.0                                                        
    datos = serie if orden == 'asc' else list(reversed(serie))         
    for x in datos:                                                    
        total = cortardigitos(total + cortardigitos(x))
    return total                                                       

def mostrar_errores(valor_real, valor_aprox):
    error_abs = abs(valor_real - valor_aprox)
    error_rel = error_abs / abs(valor_real)
    print(f"Error absoluto: |{valor_real:.9f} - {valor_aprox:.9f}| = {error_abs:.9f}")
    print(f"Error relativo: |({valor_real:.9f} - {valor_aprox:.9f}) / {valor_real:.9f}| = {error_rel:.9f}\n")

sumatoria_a = [1 / (i**2) for i in range(1, 11)]
sumatoria_b = [1 / (i**3) for i in range(1, 11)]

a_asc = sumacorte(sumatoria_a, orden='asc')
a_desc = sumacorte(sumatoria_a, orden='desc')
b_asc = sumacorte(sumatoria_b, orden='asc')
b_desc = sumacorte(sumatoria_b, orden='desc')

suma_real_a = sum(1 / (i**2) for i in range(1, 11)) 
suma_real_b = sum(1 / (i**3) for i in range(1, 11))

print("Para el literal a\n")
print(f"Sumatoria sin corte: {suma_real_a:.9f}")
print(f"Suma ascendente con corte: {a_asc}")
mostrar_errores(suma_real_a, a_asc)
print(f"Suma descendente con corte: {a_desc}")
mostrar_errores(suma_real_a, a_desc)

print("Para el literal b\n")
print(f"Sumatoria sin corte: {suma_real_b:.9f}")
print(f"Suma ascendente con corte: {b_asc}")
mostrar_errores(suma_real_b, b_asc)
print(f"Suma descendente con corte: {b_desc}")
mostrar_errores(suma_real_b, b_desc)

Para el literal a

Sumatoria sin corte: 1.549767731
Suma ascendente con corte: 1.53
Error absoluto: |1.549767731 - 1.530000000| = 0.019767731
Error relativo: |(1.549767731 - 1.530000000) / 1.549767731| = 0.012755286

Suma descendente con corte: 1.54
Error absoluto: |1.549767731 - 1.540000000| = 0.009767731
Error relativo: |(1.549767731 - 1.540000000) / 1.549767731| = 0.006302707

Para el literal b

Sumatoria sin corte: 1.197531986
Suma ascendente con corte: 1.16
Error absoluto: |1.197531986 - 1.160000000| = 0.037531986
Error relativo: |(1.197531986 - 1.160000000) / 1.197531986| = 0.031341113

Suma descendente con corte: 1.19
Error absoluto: |1.197531986 - 1.190000000| = 0.007531986
Error relativo: |(1.197531986 - 1.190000000) / 1.197531986| = 0.006289590



------

# Ejercicio 2

La serie de Maclaurin para la función arcotangente converge para −1<𝑥≤1 y está dada por:

$\arctan x = \lim_{n \to \infty} P_n(x) \equiv \lim_{n \to \infty} \sum_{i=1}^{n} (-1)^{i+1} \frac{x^{2i-1}}{2i - 1}$


a. Utilice el hecho de que tan 𝜋 4⁄ =1 para determinar el número n de términos de la serie que se necesita 
sumar para garantizar que |4𝑃𝑛(1)−𝜋|<10−3 

b. El lenguaje de programación C++ requiere que el valor de 𝜋 se encuentre dentro de 10−10. ¿Cuántos 
términos de la serie se necesitarían sumar para obtener este grado de precisión? 

In [5]:
import math

def calcularnecesario(tolerancia_error_pi):          
    # n se calcula basándose en la serie de Leibniz para pi/4
    n = math.ceil((4 / tolerancia_error_pi - 1) / 2)    
    return int(n)                                     

def calcularpi(n_terminos):                 
    suma_parcial_pi_4 = 0.0                             

    for i in range(1, n_terminos + 1):                  
        denominador = 2 * i - 1                         
        signo = (-1)**(i + 1)                           
        termino = signo / denominador                   
        suma_parcial_pi_4 += termino                    
        
    return 4 * suma_parcial_pi_4                      

pi_referencia = 3.141592653589793  

# --- Literal a ---
print("Literal a")
tolerancia_a = 10**-3  # PRIMERO definimos la variable
n_a = calcularnecesario(tolerancia_a) # LUEGO la usamos
pi_aprox_a = calcularpi(n_a)                    
error_a = abs(pi_aprox_a - pi_referencia)

print(f"Número de términos requeridos (n): {n_a}")
print(f"Pi aproximado con {n_a} términos: {pi_aprox_a}")
print(f"Error absoluto: {error_a:.6f} (< {tolerancia_a})")

# --- Literal b ---
print("\nLiteral b")
tolerancia_b = 10**-10  
n_b = calcularnecesario(tolerancia_b)
print(f"Número de términos requeridos (n): {n_b:,}")

Literal a
Número de términos requeridos (n): 2000
Pi aproximado con 2000 términos: 3.1410926536210413
Error absoluto: 0.000500 (< 0.001)

Literal b
Número de términos requeridos (n): 20,000,000,000


----

# Ejercicio 3

 
Otra fórmula para calcular 𝜋 se puede deducir a partir de la identidad 𝜋 4⁄ =4arctan1
5
−arctan 1
239
. 
Determine el número de términos que se deben sumar para garantizar una aproximación 𝜋 dentro de 10−3. 

In [6]:
import math

error_máximo = 1e-3 

x1 = 1 / 5
x2 = 1 / 239

pi_aprox = 0
n = 0                       
termino = error_máximo

while abs(termino) >= error_máximo:                     
    arctan1 = ((-1)**n) * (x1**(2*n + 1)) / (2*n + 1)
    arctan2 = ((-1)**n) * (x2**(2*n + 1)) / (2*n + 1)
    termino = 4 * arctan1 - arctan2                     
    pi_aprox += termino                                 
    n += 1                                              

print("Numero de termino:",n)

Numero de termino: 3


-----

# Ejercicio 4

 
Compare los siguientes tres algoritmos. ¿Cuándo es correcto el algoritmo de la parte 1a?

a.- ENTRADA 𝑛,𝑥1,𝑥2,⋯,𝑥𝑛. 

SALIDA PRODUCT. 

Paso 1 Determine PRODUCT = 0.

Paso 2 Para 𝑖 = 1,2,⋯,𝑛 haga 

Determine PRODUCT = PRODUCT∗𝑥𝑖. 

Paso 3 SALIDA PRODUCT; 

PARE. 

In [8]:
def algoritmo_1(lista_x):
    PRODUCT = 0  
    for x in lista_x:
        PRODUCT = PRODUCT * x
    return PRODUCT

lista_1 = [3, 8, 2]  
lista_2 = [1, 2, 0]  

print(f"Algoritmo 1 con lista de terminos diferentes de cero: {algoritmo_1(lista_1)}")
print(f"Algoritmo 1 con termino 0 en una lista: {algoritmo_1(lista_2)}")

Algoritmo 1 con lista de terminos diferentes de cero: 0
Algoritmo 1 con termino 0 en una lista: 0


b.-  ENTRADA 𝑛,𝑥1,𝑥2,⋯,𝑥𝑛. 

SALIDA PRODUCT.

Paso 1 Determine PRODUCT = 1. 

Paso 2 Para 𝑖 = 1,2,⋯,𝑛 haga 

Set PRODUCT = PRODUCT∗𝑥𝑖. 

Paso 3 SALIDA PRODUCT; 

PARE. 

In [9]:
def algoritmo_2(lista_x):
    PRODUCT = 1  
    for x in lista_x:
        PRODUCT = PRODUCT * x
    return PRODUCT

lista_1 = [3, 8, 2]  
lista_2 = [1, 2, 0]  

print(f"Algoritmo 2 con lista de términos diferentes de cero: {algoritmo_2(lista_1)}")
print(f"Algoritmo 2 con término 0 en una lista: {algoritmo_2(lista_2)}")

Algoritmo 2 con lista de términos diferentes de cero: 48
Algoritmo 2 con término 0 en una lista: 0


3.-  ENTRADA 𝑛,𝑥1,𝑥2,⋯,𝑥𝑛. 

SALIDA PRODUCT. 

Paso 1 Determine PRODUCT = 1. 

Paso 2 Para 𝑖 = 1,2,⋯,𝑛 haga 

si 𝑥𝑖=0 entonces determine PRODUCT = 0; 

SALIDA PRODUCT; 

PARE 

Determine PRODUCT = PRODUCT∗𝑥𝑖. 

Paso 3 SALIDA PRODUCT; 

PARE. 

In [10]:
def algoritmo_3(x_list):
    """Algoritmo correcto con optimización para cero."""
    PRODUCT = 1
    for x in x_list:
        if x == 0:
            PRODUCT = 0
            return PRODUCT 
        PRODUCT = PRODUCT * x
    return PRODUCT

lista_1 = [3, 8, 2]  
lista_2 = [1, 2, 0]  

print(f"Algoritmo 2 con lista de términos diferentes de cero: {algoritmo_3(lista_1)}")
print(f"Algoritmo 2 con término 0 en una lista: {algoritmo_3(lista_2)}")

Algoritmo 2 con lista de términos diferentes de cero: 48
Algoritmo 2 con término 0 en una lista: 0


------


# Ejercicio 5


a.- ¿Cuántas multiplicaciones y sumas se requieren para determinar una suma de la forma 

$$\sum_{i=1}^{n} \sum_{j=1}^{i} a_ib_j$$

b.- Modifique la suma en la parte a) a un formato equivalente que reduzca el número de cálculos. 

In [11]:
def contar_operacion(n):
    multiplicacion = n * (n + 1) // 2
    suma = n * (n - 1) // 2
    return multiplicacion, suma

n = 9
multiplicacion, suma = contar_operacion(n)
print("\nLiteral a\n")
print(f"Para n = {n}, se necesitan {multiplicacion} multiplicaciones y {suma} sumas.")

def suma_optimizada(a, b):
    n = len(a)
    suma_acumulada = [0] * n
    suma_acumulada[-1] = a[-1]

    for i in range(n - 2, -1, -1):                              
        suma_acumulada[i] = suma_acumulada[i + 1] + a[i]

    resultado = 0
    for j in range(n):                              
        resultado += b[j] * suma_acumulada[j]

    return resultado

a = [2, 4, 6, 8, 10]
b = [1, 3, 5, 7, 9]
resultado = suma_optimizada(a, b)
print("\nLiteral b\n")  
print("Resultado de la suma optimizada:", resultado)



Literal a

Para n = 9, se necesitan 45 multiplicaciones y 36 sumas.

Literal b

Resultado de la suma optimizada: 450


-----


# DISCUSIONES

# Primera Discusion

Escriba un algoritmo para sumar la serie finita

$\sum_{i=1}^{n} x_i$

𝑖=1 en orden inverso. 

```
ENTRADA: n, x₁, x₂, ..., xₙ.
SALIDA: SUMA.

Paso 1: Determine SUMA = 0.
Paso 2: Para i = n, n−1, ..., 1 haga
            Determine SUMA = SUMA + xᵢ.
Paso 3: SALIDA SUMA;
        PARE
FIN
```

-------

# Segunda Discusion

Las ecuaciones (1.2) y (1.3) en la sección 1.2 proporcionan formas alternativas para las raíces $𝑥_1$ y $𝑥_2$ de $𝑎𝑥^2 + 𝑏𝑥 + 𝑐 = 0$. Construya un algoritmo con entrada 𝑎, 𝑏, 𝑐 c y salida $𝑥_1, 𝑥_2$ que calcule las raíces $𝑥_1$ y $𝑥_2$ (que pueden ser iguales con conjugados complejos) mediante la mejor fórmula para cada raíz.

```
ENTRADA: a, b, c.
SALIDA: x₁, x₂.

Paso 1: Determine el discriminante = b² − 4ac.
Paso 2: Si discriminante < 0 entonces
            Determine parte real = −b / (2a).
            Determine parte imaginaria = sqrt(−discriminante) / (2a).
                    x₁ = parte real + parte imaginaria * i.
                    x₂ = parte real − parte imaginaria * i.
        Sino
            Determine sqrt_d = sqrt(discriminante).
            Si b ≥ 0 entonces
                x₁ = (−b − sqrt_d) / (2a).
                x₂ = (2c) / (−b − sqrt_d).
            Sino
                x₁ = (−b + sqrt_d) / (2a).
                x₂ = (2c) / (−b + sqrt_d).
            Fin Si
        Fin Si
Paso 3: SALIDA x₁, x₂;
        PARE
FIN
```

---------

# Tercera Discusion

Suponga que:

$$\frac{1−2𝑥}{1−𝑥+𝑥^2} + \frac{2𝑥−4𝑥^3}{1−𝑥^2+𝑥^4} + \frac{4𝑥^3 − 8𝑥^7}{1 − 𝑥^4 + 𝑥^8} + ⋯ = \frac{1+2𝑥}{1+𝑥+𝑥^2} ,$$


para 𝑥<1 y si 𝑥=0.25. Escriba y ejecute un algoritmo que determine el número de términos necesarios en 
el lado izquierdo de la ecuación de tal forma que el lado izquierdo difiera del lado derecho en menos de 10−6.

```
ENTRADA: x = 0.25.
SALIDA: k 

Paso 1: Determine tolerancia = 1e-6.
Paso 2: Calcule lado_derecho = (1 + 2x) / (1 + x + x²).
Paso 3: Determine suma_lado_izquierdo = 0.
Paso 4: Determine k = 0.
Paso 5: Repita
            Determine numerador = 2^k * x^(2^k − 1) − 2^(k+1) * x^(2^(k+1) − 1).
            Determine denominador = 1 − x^(2^k) + x^(2^(k+1)).
            Determine término = numerador / denominador.
            Determine suma_lado_izquierdo = suma_lado_izquierdo + término.
            Determine k = k + 1.
        Hasta que |suma_lado_izquierdo − lado_derecho| < tolerancia.
Paso 6: SALIDA k;
        PARE
FIN
```